portal-impl/src/com/liferay/portal/upgrade/v7_0_3/UpgradePermission.java

```diff
@@ -0,0 +1,268 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.upgrade.v7_0_3;
+
+import com.liferay.portal.kernel.security.permission.ResourceActionsUtil;
+import com.liferay.portal.kernel.upgrade.UpgradeException;
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+import com.liferay.portal.kernel.util.LoggingTimer;
+import com.liferay.portal.kernel.util.StringBundler;
+import com.liferay.portal.util.PropsValues;
+import com.liferay.portlet.asset.service.permission.AssetPermission;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author Pavel Savinov
+ */
+public class UpgradePermission extends UpgradeProcess {
+
+	protected int checkActions(List<Integer> resourceActions, int actionsIds) {
+		int newActionsIds = 0;
+
+		for (Integer resourceAction : resourceActions) {
+			if ((actionsIds & resourceAction.intValue()) != 0) {
+				newActionsIds += resourceAction.intValue();
+			}
+		}
+
+		return newActionsIds;
+	}
+
+	protected int checkCategoriesActions(int actionsIds) throws Exception {
+		List<Integer> resourceActions = getResourceActions(
+			AssetPermission.RESOURCE_NAME_CATEGORIES);
+
+		return checkActions(resourceActions, actionsIds);
+	}
+
+	protected int checkTagsActions(int actionsIds) throws Exception {
+		List<Integer> resourceActions = getResourceActions(
+			AssetPermission.RESOURCE_NAME_TAGS);
+
+		return checkActions(resourceActions, actionsIds);
+	}
+
+	protected void deleteResourceActions() throws UpgradeException {
+		try (LoggingTimer loggingTimer = new LoggingTimer()) {
+			StringBundler sb = new StringBundler(3);
+
+			sb.append("delete from ResourceAction where name = '");
+			sb.append(AssetPermission.RESOURCE_NAME);
+			sb.append("';");
+
+			runSQL(sb.toString());
+		}
+		catch (Exception e) {
+			throw new UpgradeException(e);
+		}
+	}
+
+	@Override
+	protected void doUpgrade() throws Exception {
```

Comment:
I didn't have time to do an in-depth review of this class looking for potential bugs, but conceptually it makes sense to me.